rfcs/20231025-reusable-kernels/README.md

```diff
@@ -0,0 +1,106 @@
+# Proposal to add a reusable kernel knob
+
+## Introduction
+
+Primitive creation is an expensive operation, which is why oneDNN has a
+primitive cache, a persistent cache, and a kernel cache (for the GPU). For GPU
+implementations, the kernel cache is optional and must be opted into. Although
+compiling kernels for a given problem shape can open up optimization potential,
+it comes at the cost of being incompatible with the kernel cache because the
+kernel cannot be reused for different problem shapes. Therefore, it can be
+helpful to the user to be able to specify if they prefer to use a kernel that
+has been designed to be reusable via the kernel cache (accepting any possible
+performance impacts associated with it) or the typical performance-optimized
+kernel. This proposal is to add a knob the user can control to specify whether
```

Comment:
Can you please elaborate on how/when you expect the end user to use this feature? Do frameworks plan to expose it?  Have you discussed it with the framework teams?

Reply:
We have been in discussions with frameworks regarding this feature, which is why I mentioned that option 4 (primitive attribute) is preferred by frameworks. To my knowledge, they don't plan on exposing this option to the user. The idea is to use this knob for models which contain a large number of unique (dynamic) shapes/primitives to give a hint to oneDNN to use flexible/reusable kernels and reduce primitive creation time as much as possible.

Reply:
@Simonsays095, this solution sounds like something we could do in IPEX, but not in upstream. Did we discuss this with Tensorflow and Pytorch developers?

Reply:
> The idea is to use this knob for models which contain a large number of unique (dynamic) shapes/primitives to give a hint to oneDNN to use flexible/reusable kernels and reduce primitive creation time as much as possible.

From what I recall of the email chain, it seems FWKs cannot know how dimensions will change between two executions, so FWK/IPEX cannot know when to set the attribute or not. My understanding is that they would always set the hint which would lead to performance regressions when shapes do not effectively change from one run to the next.

Reply:
> My understanding is that they would always set the hint which would lead to performance regressions when shapes do not effectively change from one run to the next.

I think this (enabling this hind by default) is not an option. We expect the users to opt in to use reusable kernels in \*some\* scenarios when it makes sense to do so (e.g. for dynamic-size models).